### ISSUE TITLE
hmr? (hot module reloading)

### ISSUE THREAD

**Issue by jacoobes** [sern-handler/cli#62]
Most clis enable users with a watch command
- rollup
- eleventy
- typescript
- nodemon
- vite
Letting sern deal with this would be very convenient for users. 
We could also just include it in the future with global plugins, once compatibility branch rolls around

**Comment by Murtatrxx** [comment]
I'm not sure @EvolutionX-10 is here

**Comment by Murtatrxx** [comment]
He's taking his long break I think

**Comment by Murtatrxx** [comment]
Can we do it without him?

**Comment by jacoobes** [comment]
need his permission as he's author.

**Comment by Murtatrxx** [comment]
🤔 weren't you the owner?

**Comment by EvolutionX-10** [comment]
> Most clis enable users with a watch command
> 
> * rollup
> * eleventy
> * typescript
> * nodemon
> * vite
>   Letting sern deal with this would be very convenient for users.
>   We could also just include it in the future with global plugins, once compatibility branch rolls around

`sern watch` ?
what things will it do 🤔

**Comment by jacoobes** [comment]
Unlike nodemon it wouldn’t restart. But nodemon also works idk

**Comment by jacoobes** [comment]
I’ve messed around with hmr with vite and honestly it’s one of the most
productive things I’ve used

**Comment by EvolutionX-10** [comment]
Yep I've used hmr as well in a discord bot before, I loved it
However it doesn't work good with TS if we want no restarts since TS compiling to JS is done as a whole project not a single file itself

**Comment by jacoobes** [comment]
Yep, true, could tell typescript users to use the `tsc watch` command however and then hmr on distribution

**Comment by EvolutionX-10** [comment]
Also check #63
had in my mind from a long time and finally completed it, if you got more improvements, you can commit there and notify me

**Comment by EvolutionX-10** [comment]
If you want you can create a branch and work on it, else I'll work on it by next week
My account will be enabled soon btw, exam's this week

**Comment by EvolutionX-10** [comment]
```bash
git fetch origin
git checkout feat/hmr
```

**Comment by jacoobes** [comment]
POG, looking forward to return
Hmr experimenting should be good

**Comment by vKxni** [comment]
Ok cool guys

**Comment by jacoobes** [comment]
🥸

**Comment by jacoobes** [comment]
Will be working on this. Experimental at the moment! A dev i saw ( Chooksie ) has a really good explanation on this, so I will research into it.

![image](https://user-images.githubusercontent.com/76754747/180873787-56480cab-99d2-4272-82d6-668553ff9fe7.png)

**Comment by jacoobes** [comment]
Well just discovered hmr and nodemon near the same thing. nodemon is more for server side stuff while hmr is frontend.
https://stackoverflow.com/questions/46086297/nodemon-and-or-hot-reloading-with-a-node-react-web-app. I suppose we can just do some chokidar based system? @leonardobagi recommended this can be done in the framework and not in the cli

**Comment by leonardobagi** [comment]
> I suppose we can just do some chokidar based system? @leonardobagi recommended this can be done in the framework and not in the cli

This feature could either be directly implemented in the `Sern.init` command via an option, or made via plugin. I do think that making built-in in Sern can be lighter, faster (since there's no need for external packages) and more customizable; it could also be easier to implement, because controlling command files via plugin can generate conflicts with the main code. The only drawback, though, is that we couldn't watch for an X file rather than the entire folder, but that is not such a big deal

**Comment by leonardobagi** [comment]
A `sern watch` cli command wouldn't be that good imo, since the init thing of the cli only generates the code

**Comment by jacoobes** [comment]
> A `sern watch` cli command wouldn't be that good imo, since the init thing of the cli only generates the code
Understandable.
I was thinking these ways to implement a module watcher

- through cli
- through plugins
- through global plugins ( experimental (and top secret!!!! jk) )
- built in

one part of Sern i want to keep is to keep customization and to not add any extra features that users may not want.
So with this reasoning, I would say global plugins or built in is the best option. Like you said here :


> > I suppose we can just do some chokidar based system? @leonardobagi recommended this can be done in the framework and not in the cli
> 
> This feature could either be directly implemented in the `Sern.init` command via an option, or made via plugin. I do think that making built-in in Sern can be lighter, faster (since there's no need for external packages) and more customizable; it could also be easier to implement, because controlling command files via plugin can generate conflicts with the main code. The only drawback, though, is that we couldn't watch for an X file rather than the entire folder, but that is not such a big deal

Individual command watchers is just inefficient and id rather have a whole directory being watched.
So that brings us down to two options

- global plugins 
- built in
Now the question boils down to is adding an hmr system ( in return adding more to the depedencies list most likely ) built into the framework wise? Or would it best be handled via the user, to opt into it through plugins.

**Comment by leonardobagi** [comment]
Now that you mentioned global plug-ins thing (which is probably what I wanted to do), it's better to apply that option.

It's very likely some users will prefer nodemon over this watcher, since, sometimes, the user needs to hot reload other files outside the commands directory (a utils folder e.g).

In addition, making it a plugin will reduce unnecessary and unwanted dependencies. (Because chokidar would be required, and some users will just not use the feature, increasing final bundle size)

**Comment by leonardobagi** [comment]
I also think that promoting plug-ins with that hmr one will make them look for more utilities plug-ins, create their own and contribute to existing ones. It's a nice way to a promote such an unique feature in handlers.

**Comment by leonardobagi** [comment]
> It's very likely some users will prefer nodemon over this watcher, since, sometimes, the user needs to hot reload other files outside the commands directory (a utils folder e.g).

Our main point is to avoid multiple bot relogins and restarts, and only dynamically reload the commands, making everything faster and using less bandwidth. We can even make a dev slash command `/reload`, if it fits here.

**Comment by SrIzan10** [comment]
reviving this, I want a convenient way of developing my bot without using ts-node or nodemon!

**Comment by leonardobagi** [comment]
It seems like the plug-ins API can't get us the commands yet, so we'll have to wait for that to happen

**Comment by SrIzan10** [comment]
yea some guy at the [discord](https://sern-handler.js.org/discord) just asked for that

**Comment by leonardobagi** [comment]
Yea I did

**Comment by SrIzan10** [comment]
oh wait its you?

**Comment by SrIzan10** [comment]
oh yeah you are 💀

**Comment by leonardobagi** [comment]
Lmfao

**Comment by jacoobes** [comment]
@leonardobagi  has been working on an experimental FileWatcher that may work

**Comment by leonardobagi** [comment]
That class starts the whole hmr thing, the only thing remaining is syncing the commands on discord. This could be easily done via (https://github.com/sern-handler/sern-community/pull/30)[a CommandsSyncer] class I made for the sern-community bot.
